src/libiotc/iotc_roughtime_client.cc

```diff
@@ -0,0 +1,225 @@
+  socket_evts[0].in_socket_want_read = 1;
+  bool ready_to_read = false;
+  while (!ready_to_read) {
+    state =
+        iotc_bsp_io_net_select(socket_evts, socket_evts_size, kTimeoutSeconds);
+    switch (state) {
+    case IOTC_BSP_IO_NET_STATE_OK:
+      if (socket_evts[0].out_socket_can_read == 1) {
+        ready_to_read = true;
+      }
+      break;
+    case IOTC_BSP_IO_NET_STATE_TIMEOUT:
+      return IOTC_ROUGHTIME_TIMEOUT_ERROR;
+    case IOTC_BSP_IO_NET_STATE_ERROR:
+      return IOTC_ROUGHTIME_NETWORK_ERROR;
+    default:
+      break;
+    }
+  }
+
+  uint8_t recv_buf[roughtime::kMinRequestSize];
+  int buf_len;
+  state = iotc_bsp_io_net_read(socket, &buf_len, recv_buf, sizeof(recv_buf));
+  if (IOTC_BSP_IO_NET_STATE_ERROR == state) {
+    iotc_debug_logger("Read from the socket");
+    iotc_bsp_io_net_close_socket(reinterpret_cast<iotc_bsp_socket_t*>(&socket));
+    return IOTC_ROUGHTIME_NETWORK_ERROR;
+  }
+  const uint64_t kEndUs = iotc_bsp_time_getmonotonictime_milliseconds();
+
+  iotc_bsp_io_net_close_socket(reinterpret_cast<iotc_bsp_socket_t*>(&socket));
+
+  if (buf_len == -1) {
+    if (errno == EINTR) {
+      iotc_debug_format(stderr, "No response from %s with %d seconds.\n",
+                        server_name.c_str(), kTimeoutSeconds);
+      return IOTC_ROUGHTIME_TIMEOUT_ERROR;
+    }
+
+    iotc_debug_logger("Read from the socket");
+    return IOTC_ROUGHTIME_NETWORK_ERROR;
+  }
+
+  std::string error;
+  if (!roughtime::ParseResponse(
+          &(time_data->timestamp), &(time_data->radius), &error,
+          reinterpret_cast<const uint8_t*>(server_public_key.data()), recv_buf,
+          buf_len, nonce)) {
+    iotc_debug_format(stderr, "Response from %s failed verification: %s",
+                      server_name.c_str(), error.c_str());
+    return IOTC_ROUGHTIME_RECEIVE_TIME_ERROR;
+  }
+
+  // We assume that the path to the Roughtime server is symmetric and thus
+  // add half the round-trip time to the server's timestamp to produce our
+  // estimate of the current time.
+  time_data->reply_time = (kEndUs - kStartUs);
+  time_data->timestamp /= 1000;
+  time_data->timestamp += (time_data->reply_time) / 2;
+
+  return IOTC_ROUGHTIME_OK;
+}
+
+iotc_roughtime_state_t
+iotc_roughtime_getcurrenttime(const char* name, const char* public_key,
+                              const char* server_address,
+                              iotc_roughtime_timedata_t* time_data) {
+  int socket;
+  iotc_roughtime_state_t state;
+  for (int i = 0; i < kRetrialTimes; i++) {
+    if ((state = iotc_roughtime_create_socket(&socket, server_address)) !=
+        IOTC_ROUGHTIME_OK) {
+      iotc_debug_logger("Failed on connecting socket.");
+      continue;
+    }
+    if ((state = iotc_roughtime_receive_time(socket, name, public_key,
+                                             time_data)) != IOTC_ROUGHTIME_OK) {
+      iotc_debug_logger("Failed on receving time is failed.");
+      continue;
```

Comment:
socket should be closed here.

Reply:
Already has some closing statement in receive_time(): (before returning error) + (Close socket after reading from it). 
To prevent socket closed twice, I just added missing closing statement before returning error in receive_time(). Tell me if there is better sol!

Reply:
Mhm, I see. What is bugging me here is the symmetric responsibility of socket creation and closure. I'd assume the same entity is responsible for socket creation and closure. E.g. the `_receive_time` receives a functional socket but returns a closed socket assuming it won't be used further.
I'd remove socket close responsibility from `_receive_time` function and give it to the one which creates the socket. Thus the lifetime/functionality of socket is controlled by a single entity, and furthermore - maybe won't ever happen, but still - the socket can be used even after `_receive_time` function has finished.